src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -807,9 +809,10 @@ namespace cryptonote
         return 1;
       }
       //to avoid concurrency in core between connections, suspend connections which delivered block later then first one
+      const crypto::hash block_hash = get_block_hash(b);
       if(count == 2)
       {
-        if(m_core.have_block(get_block_hash(b)))
+        if(m_core.have_block(block_hash))
```

Comment:
I was clearly hoping that there was a quick way for a thread to return without having to wait for the lock. But if it passes the `.back()` check I proposed, then there is a good chance it will not have to wait at the lock (because the other thread has finished updating). The only beneficial scenario for an unlocked `.back()` check seems to be if 2 responses arrive after a prior response previously updated the same blocks. Both threads could return immediately instead of one having to wait for the other to do its redundancy checks. But even then, the expected wait should be short.

Have you considered removing this check entirely? This is an additional synchronization point that can be removed (assuming we both understand this block of code correctly).